File: cpp/RoundTrip/src/ping.cpp
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This is a simple throughput application measuring obtainable throughput using the thingSDK
 */
#include <chrono>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdlib.h>
#include <thread>
#include <time.h>
#ifdef _WIN32
#include <Windows.h>
#endif

#include <IoTDataThing.hpp>
#include <JSonThingAPI.hpp>
#include <ThingAPIException.hpp>

#include "include/cxxopts.h"
#include "include/utilities.h"

using namespace std;
using namespace com::adlinktech::datariver;
using namespace com::adlinktech::iot;

static ExampleTimeStats roundTripOverall = exampleInitTimeStats();
static ExampleTimeStats writeAccessOverall = exampleInitTimeStats();
static ExampleTimeStats readAccessOverall = exampleInitTimeStats();

static void showStats(bool overall, unsigned long elapsedSeconds, ExampleTimeStats roundTrip, ExampleTimeStats writeAccess, ExampleTimeStats readAccess) {
	if (overall) {
		cout << endl << COLOR_GREEN << "# Overall";
	} else {
		cout << setw(9) << right << elapsedSeconds;
	}

	cout
		<< setw(10) << right << roundTrip.values.size()
		<< setw(9) << right << fixed << setprecision(0) << exampleGetMedianFromTimeStats(roundTrip)
		<< setw(9) << right << roundTrip.min
		<< setw(11) << right << writeAccess.values.size()
		<< setw(9) << right << fixed << setprecision(0) << exampleGetMedianFromTimeStats(writeAccess)
		<< setw(9) << right << writeAccess.min
		<< setw(11) << right << readAccess.values.size()
		<< setw(9) << right << fixed << setprecision(0) << exampleGetMedianFromTimeStats(readAccess)
		<< setw(9) << right << readAccess.min << NO_COLOR << endl;
}

#ifdef _WIN32
static bool ctrlHandler(DWORD fdwCtrlType)
{
    showStats(true, 0, roundTripOverall, writeAccessOverall, readAccessOverall);
    resetConsoleMode();
	ExitProcess(1); // perform a brute-force exit for now
    return true;
}
#else
static void ctrlHandler(int fdwCtrlType)
{
    showStats(true, 0, roundTripOverall, writeAccessOverall, readAccessOverall);
    exit(1);
}
#endif

class Ping {
private:
    string m_thingPropertiesUri;
    DataRiver m_dataRiver = createDataRiver();
    Thing m_thing = createThing();
	IOT_NVP_SEQ m_sampleData;

    DataRiver createDataRiver() {
        return DataRiver::getInstance();
    }

    Thing createThing() {
        // Create and Populate the TagGroup registry with JSON resource files.
        JSonTagGroupRegistry tgr;
        tgr.registerTagGroupsFromURI("file://definitions/TagGroup/com.adlinktech.example/PingTagGroup.json");
        tgr.registerTagGroupsFromURI("file://definitions/TagGroup/com.adlinktech.example/PongTagGroup.json");
        m_dataRiver.addTagGroupRegistry(tgr);

        // Create and Populate the ThingClass registry with JSON resource files.
        JSonThingClassRegistry tcr;
        tcr.registerThingClassesFromURI("file://definitions/ThingClass/com.adlinktech.example/PingThingClass.json");
        m_dataRiver.addThingClassRegistry(tcr);

        // Create a Thing based on properties specified in a JSON resource file.
        JSonThingProperties tp;
        tp.readPropertiesFromURI(m_thingPropertiesUri);

        return m_dataRiver.createThing(tp);
    }

	void waitForPong() {
		// wait for pong to appear by discovering its thingId and thingClass
		cout << "# Waiting for pong to run..." << endl;
		auto discoveredThingRegistry = m_dataRiver.getDiscoveredThingRegistry();
		bool readerFound = false;

		while (!readerFound) {
			try {
				// see if we already know pongs's thing class
				auto thing = discoveredThingRegistry.findDiscoveredThing("pongThing1", "Pong:com.adlinktech.example:v1.0");
				readerFound = true;
			} catch (...) {}
			this_thread::sleep_for(chrono::seconds(1));
		}
	}

	void initPayload(unsigned long payloadSize) {
	    IOT_VALUE payload;
		payload.iotv_byte_seq(IOT_BYTE_SEQ());
		for (unsigned long i = 0; i < payloadSize; i++) {
			payload.iotv_byte_seq().push_back('a');
		}
		m_sampleData = { IOT_NVP(string("payload"), payload) };
	}

	void warmUp() {
		Timepoint startTime = Clock::now();
		int64_t waitTimeout = 10000;

		cout << "# Warming up 5s to stabilise performance..." << endl;
		while (Duration<Microseconds>(Clock::now() - startTime) / US_IN_ONE_SEC < 5) {
			m_thing.write("Ping", m_sampleData);
			m_thing.read<IOT_NVP_SEQ>("Pong", waitTimeout);
		}
		cout << "# Warm up complete" << endl;
	}



public:
    Ping(string thingPropertiesUri) :
        m_thingPropertiesUri(thingPropertiesUri)
    {
        cout << "# Ping started" << endl;
    }

    ~Ping() {
        m_dataRiver.close();
        cout << "# Ping stopped" << endl;
    }

	int sendTerminate() {
        cout << "# Sending termination request." << endl;
        m_thing.purge("Ping", "ping");
        this_thread::sleep_for(chrono::seconds(1));
		return 0;
	}

    int run(unsigned long payloadSize, unsigned long numSamples, unsigned long runningTime) {
		Timepoint startTime;
		Timepoint preWriteTime;
		Timepoint postWriteTime;
		Timepoint preReadTime;
		Timepoint postReadTime;
		int64_t waitTimeout = 10000;
		ExampleTimeStats roundTrip = exampleInitTimeStats();
		ExampleTimeStats writeAccess = exampleInitTimeStats();
		ExampleTimeStats readAccess = exampleInitTimeStats();

		cout << "# Parameters: payload size: " << payloadSize << " | number of samples: " << numSamples << " | running time: " << runningTime << endl;

		// Wait for the Pong Thing
		waitForPong();

		// Init payload
		initPayload(payloadSize);

		// Warm-up for 5s
		warmUp();

		cout << "# Round trip measurements (in us)" << endl;
		cout << COLOR_LMAGENTA << "#             Round trip time [us]         Write-access time [us]       Read-access time [us]" << NO_COLOR << endl;
		cout << COLOR_LMAGENTA << "# Seconds     Count   median      min      Count   median      min      Count   median      min" << NO_COLOR << endl;

		startTime = Clock::now();
		unsigned long elapsedSeconds = 0;
		for (unsigned long i = 0; !numSamples || i < numSamples; i++) {
			// Write a sample that pong can send back
			preWriteTime = Clock::now();
			m_thing.write("Ping", m_sampleData);
			postWriteTime = Clock::now();

			// Read sample
			preReadTime = Clock::now();
			vector<DataSample<IOT_NVP_SEQ>> samples = m_thing.read<IOT_NVP_SEQ>("Pong", waitTimeout);
			postReadTime = Clock::now();

			// Validate sample count
			if (samples.size() != 1) {
				cerr << "ERROR: Ping received " << samples.size() << " samples but was expecting 1." << endl;
				return 1;
			}

			// Update stats
			writeAccess += (unsigned long)Duration<Microseconds>(postWriteTime - preWriteTime);
			readAccess += (unsigned long)Duration<Microseconds>(postReadTime - preReadTime);
			roundTrip += (unsigned long)Duration<Microseconds>(postReadTime - preWriteTime);
			writeAccessOverall += (unsigned long)Duration<Microseconds>(postWriteTime - preWriteTime);
			readAccessOverall += (unsigned long)Duration<Microseconds>(postReadTime - preReadTime);
			roundTripOverall += (unsigned long)Duration<Microseconds>(postReadTime - preWriteTime);

			// Print stats each second
			if ((Duration<Microseconds>(postReadTime - startTime) > US_IN_ONE_SEC) || (i && i == numSamples)) {
				// Print stats
				showStats(false, ++elapsedSeconds, roundTrip, writeAccess, readAccess);

				// Reset stats for next run
				exampleResetTimeStats(roundTrip);
				exampleResetTimeStats(writeAccess);
				exampleResetTimeStats(readAccess);

				// Set values for next run
				startTime = Clock::now();

				// Check for timeout
				if (runningTime > 0 && elapsedSeconds >= runningTime) {
					break;
				}
			}
		}

		// Print overall stats
		showStats(true, 0, roundTripOverall, writeAccessOverall, readAccessOverall);

		return 0;
	}
};


static void GetCommandLineParameters(int argc, char *argv[],
        unsigned long& payloadSize,
        unsigned long& numSamples,
        unsigned long& runningTime,
		bool& quit
) {
    cxxopts::Options options("ThroughputWriter", "ADLINK ThingSDK ThroughputWriter");
    try {
        options.add_options()
            ("p,payload-size", "Payload size", cxxopts::value<unsigned long>()->default_value("0"))
            ("n,num-samples", "Number of samples (0 is infinite)", cxxopts::value<unsigned long>()->default_value("0"))
            ("r,running-time", "Running Time in seconds (0 is infinite)", cxxopts::value<unsigned long>()->default_value("0"))
			("q,quit", "Send a quit signal to pong", cxxopts::value<bool>())
            ("h,help", "Print help")
            ;
        options.parse_positional({"payload-size", "num-samples", "running-time", "other"});
        options.positional_help("[payload-size] [num-samples] [running-time]");

        auto cmdLineOptions = options.parse(argc, argv);

        if (cmdLineOptions.count("help")) {
            cout << options.help({""}) << endl;
            exit(0);
        }

		quit = cmdLineOptions["q"].as<bool>();
        payloadSize = cmdLineOptions["p"].as<unsigned long>();
        numSamples = cmdLineOptions["n"].as<unsigned long>();
        runningTime = cmdLineOptions["r"].as<unsigned long>();
    }
    catch (cxxopts::OptionException e) {
        cerr << e.what() << endl << endl;
        cout << options.help({""}) << endl;
        exit(1);
    }
}

int main(int argc, char *argv[]) {
	unsigned long payloadSize = 0;
    unsigned long numSamples = 0;
    unsigned long runningTime = 0;
	bool quit = false;
   	GetCommandLineParameters(argc, argv, payloadSize, numSamples, runningTime, quit);

    // Register handler for Ctrl-C
    registerControlHandler();

#ifdef _WIN32
    setConsoleMode();
#endif

	int result = 0;
    try
    {
        Ping ping("file://./config/PingProperties.json");
		if (quit) {
			result = ping.sendTerminate();
		} else {
	        result = ping.run(payloadSize, numSamples, runningTime);
		}
    }
    catch (ThingAPIException e)
    {
        cerr << "An unexpected error occurred: " << e.what() << endl;
    }catch(std::exception& e1){
        cerr << "An unexpected error occurred: " << e1.what() << endl;
    }

#ifdef _WIN32
    resetConsoleMode();
#endif

    unregisterControlHandler();

	return result;
}

File: dotnet/RoundTrip/Pinger/Ping.cs
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This is a simple throughput application measuring obtainable 
 * throughput using the Edge SDK
 */

using System;
using System.Collections.Generic;
using Adlinktech.DataRiver;

namespace Pinger
{
    public class Ping : IDisposable
    {

        public static ExampleTimeStats RoundTripOverall = new ExampleTimeStats();
        public static ExampleTimeStats WriteAccessOverall = new ExampleTimeStats();
        public static ExampleTimeStats ReadAccessOverall = new ExampleTimeStats();    

        public bool Stop { get; set; }
        private DataRiver DataRiver { get; }
        private Thing Thing { get; }
        private IotNvpSeq SampleData { get; }          

        public Ping(string thingPropertiesURI)
        {
            DataRiver = DataRiver.GetInstance();
            Thing = CreateThing(DataRiver, thingPropertiesURI);
            SampleData = new IotNvpSeq();
            Stop = false;
            Console.CancelKeyPress += this.CtrlCHandler;
            Console.WriteLine("# Ping started");
        }

        public void Dispose()
        {
            DataRiver.Close();
            DataRiver.Dispose();
            Console.WriteLine("# Ping stopped");
        }

        private void CtrlCHandler(object sender, ConsoleCancelEventArgs args)
        {
            Stop = true;
            args.Cancel = true; // don't terminate the process, we'll do that.
            SendTerminate();

            ShowStats(true, 0, RoundTripOverall, WriteAccessOverall, ReadAccessOverall);

        }

        private static Thing CreateThing(in DataRiver dataRiver, string thingPropertiesURI)
        {
            // Create and Populate the TagGroup registry with JSON resource files.
            var tgr = new JSonTagGroupRegistry();
            tgr.RegisterTagGroupsFromUri("file://definitions/TagGroup/com.adlinktech.example/PingTagGroup.json");
            tgr.RegisterTagGroupsFromUri("file://definitions/TagGroup/com.adlinktech.example/PongTagGroup.json");
            dataRiver.AddTagGroupRegistry(tgr);

            // Create and Populate the ThingClass registry with JSON resource files.
            var tcr = new JSonThingClassRegistry();
            tcr.RegisterThingClassesFromUri("file://definitions/ThingClass/com.adlinktech.example/PingThingClass.json");
            dataRiver.AddThingClassRegistry(tcr);

            // Create a Thing based on properties specified in a JSON resource file.
            var tp = new JSonThingProperties();
            tp.ReadPropertiesFromUri(thingPropertiesURI);

            return dataRiver.CreateThing(tp);
        }

        public int Run(Options options)
        {
            var payloadSize = options.PayloadSize;
            var numSamples = options.NumSamples;
            var runningTime = options.RunningTime;

            DateTime startTime;
            DateTime preWriteTime;
            DateTime postWriteTime;
            DateTime preReadTime;
            DateTime postReadTime;

	    int waitTimeout = 10000;

	    ExampleTimeStats roundTrip = new ExampleTimeStats();
	    ExampleTimeStats writeAccess = new ExampleTimeStats();
	    ExampleTimeStats readAccess = new ExampleTimeStats();

	    Console.WriteLine("# Parameters: payload size: " + payloadSize 
                + " | number of samples: "  + numSamples
                + " | running time: " + runningTime);

	    // Wait for the Pong Thing
	    WaitForPong();

	    // Init payload
	    InitPayload(payloadSize);

	    // Warm-up for 5s
	    WarmUp();

	    Console.WriteLine("# Round trip measurements (in us)");

	    Console.ForegroundColor = ConsoleColor.Magenta;
            Console.WriteLine("#             Round trip time [us]         Write-access time [us]       Read-access time [us]");
            Console.WriteLine("# Seconds     Count   median      min      Count   median      min      Count   median      min");
            Console.ResetColor();

	    startTime = DateTime.Now;	    
            ulong elapsedSeconds = 0;

	    for (ulong i = 0; numSamples == 0 || i < numSamples; i++)
            {
	        preWriteTime = DateTime.Now;		
                Thing.Write("Ping", SampleData);		
                postWriteTime = DateTime.Now;

		// Read sample
		preReadTime = DateTime.Now;
                var samples = Thing.ReadIotNvp("Pong", waitTimeout);
                postReadTime = DateTime.Now;

                if (Stop)
                {
		    return 0;
                }

		// Validate sample count
		if (samples.Count != 1)
                {
                    Console.WriteLine("ERROR: Ping received " + samples.Count + " samples but was expecting 1."); 
                    return 1;
                }

		// Update stats
		var writeTimeMicroSecs = Convert.ToUInt64((postWriteTime.Subtract(preWriteTime).TotalSeconds) * 1000000);
                var readTimeMicroSecs = Convert.ToUInt64((postReadTime.Subtract(preReadTime).TotalSeconds) * 1000000);
                var roundTripMicroSecs = Convert.ToUInt64((postReadTime.Subtract(preWriteTime).TotalSeconds) * 1000000);

                writeAccess.AddStats(writeTimeMicroSecs);
                readAccess.AddStats(readTimeMicroSecs);
                roundTrip.AddStats(roundTripMicroSecs);
                WriteAccessOverall.AddStats(writeTimeMicroSecs);
                ReadAccessOverall.AddStats(readTimeMicroSecs);
                RoundTripOverall.AddStats(roundTripMicroSecs);

		// Print stats each second
		if ( (postReadTime.Subtract(startTime).TotalSeconds) > 1 ||  (i == numSamples - 1) )  
                {
		    // Print stats
		    ShowStats(false, ++elapsedSeconds, roundTrip, writeAccess, readAccess);

		    // Reset stats for next run
		    roundTrip.ResetStats();
                    writeAccess.ResetStats();
                    readAccess.ResetStats();

		    // Set values for next run
		    startTime = DateTime.Now;

		    // Check for timeout
		    if (runningTime > 0 && elapsedSeconds >= runningTime) 
                    {
		        break;
		    }
		}

	    }  

	    // Print overall stats
	    ShowStats(true, 0, RoundTripOverall, WriteAccessOverall, ReadAccessOverall);

	    return 0;
        }

        private void ShowStats(bool overall, 
            ulong elapsedSeconds, 
            ExampleTimeStats roundTrip, 
            ExampleTimeStats writeAccess, 
            ExampleTimeStats readAccess)
        {
            if (overall)
            {
                Console.ForegroundColor = ConsoleColor.DarkGreen;
                Console.Write("# Overall");
            }
            else 
            {
                Console.Write("{0, 9}", elapsedSeconds);  
            }

	     Console.Write("{0,10}", roundTrip.SortedValues.Count);
             Console.Write("{0,9}", roundTrip.CalculateMedian());
             Console.Write("{0,9}", roundTrip.Min);
             Console.Write("{0,11}", writeAccess.SortedValues.Count);
             Console.Write("{0,9}", writeAccess.CalculateMedian());
             Console.Write("{0,9}", writeAccess.Min);
             Console.Write("{0,11}", readAccess.SortedValues.Count);
             Console.Write("{0,9}", readAccess.CalculateMedian());             
             Console.WriteLine("{0,9}", readAccess.Min);
             Console.ResetColor();

        }

        private void WaitForPong() 
        {
            // wait for pong to appear by discovering its thingId and thingClass
            Console.WriteLine("# Waiting for pong to run...");
            var discoveredThingRegistry = DataRiver.DiscoveredThingRegistry;
            bool readerFound = false;

            while (!Stop && !readerFound) 
            {
                try 
                {
                    var thing = discoveredThingRegistry.FindDiscoveredThing("pongThing1", "Pong:com.adlinktech.example:v1.0");
                    readerFound = true;
                }
                catch (Exception e) 
                {
                    //Sleep for 1 second
                    System.Threading.Thread.Sleep(1000);
                }
            }
        }

        private void InitPayload(ulong payloadSize)
        {
            var payload = new IotValue();
            var byteSequence = new IotByteSeq();
            for (ulong i = 0; i < payloadSize; i++)
            {
                byte a = (byte)'a';
                byteSequence.Add(a);
            }
            payload.ByteSeq = byteSequence;
            SampleData.Add(new IotNvp("payload", payload));

        }

        private void WarmUp()
        {
            var startTime = DateTime.Now;
            var waitTimeout = 10000;

            Console.WriteLine("# Warming up 5s to stabilise performance...");
            while ( ((DateTime.Now - startTime).TotalSeconds ) < 5)
            {
                Thing.Write("Ping", SampleData);
                Thing.ReadIotNvp("Pong", waitTimeout);
            }
            Console.WriteLine("# Warm up complete");

        }

        private int SendTerminate() 
        {
            Console.WriteLine("# Sending termination request.");
            Thing.Purge("Ping", "ping");
            System.Threading.Thread.Sleep(1000);
            return 0;
        }       

    }


    public class ExampleTimeStats 
    {
        public SortedList<int, ulong> SortedValues { get; set; }
        public double Average { get; set; }   
        public ulong Min { get; set; }
        public ulong Max { get; set; }
       
        public ExampleTimeStats()
        {
            SortedValues = new SortedList<int, ulong>();
            Average = 0;
            Min = 0;
            Max = 0;            
        }

        public void AddStats(ulong microseconds)
        {                        
	    Average = ((SortedValues.Count * Average) + microseconds) / (SortedValues.Count + 1);             
            SortedValues.Add(SortedValues.Count, microseconds);

            // check min
            if (Min == 0 || microseconds < Min)
            {
                Min = microseconds;
            }
            
            // check max
            if (microseconds > Max)
            {
                Max = microseconds;
            }           
        }

        public void ResetStats()
        {
            SortedValues = new SortedList<int, ulong>();
            Average = 0;
            Min = 0;
            Max = 0; 
        }

        public ulong CalculateMedian()
        {
             var stats = SortedValues.Values;

             if (SortedValues.Count % 2 == 0) 
             {
                 ulong value1 = stats[SortedValues.Count / 2];
                 ulong value2 = stats[(SortedValues.Count / 2) + 1];   
                 return  (value1 + value2) / 2;
             }
             else 
             {
                 int medianIndex = (SortedValues.Count + 1)/ 2;
                 return stats[medianIndex];
             }
        }

    }

}
